Healthcheck

ranlu opened the pull request:
Check the health status of the container for qos

ranlu committed:
Rename variable

ranlu committed:
Check the health status of the container for qos

When the docker container implements its own health check, use it for
qos. Fallback to the cpu usage based approach only when there is no
health status entry in the state output.

nicholasturner1 commented on plugins/custom/docker_custom.py:
```diff
@@ -181,13 +181,23 @@ def _read_task_stats(self):
+                 else:
+                    unhealthy_count += 1
+                    self.log.info('Health check Failed')
+            except:
```
This is minor, but do you know what kind of exception it causes if the container doesn't have a HEALTHCHECK command? I could imagine something like the dict lookup calls raising an exception (e.g., if the api keys change) and the log message wouldn't say anything about it. It would be a graceful failure since it would just behave like the old qos check.

ranlu replied:
If the api stays the same, `inspect_container` should always return a dict, it might raise a `docker.errors.APIError` error if the docker server fails. If the container does not have a health check, the dict look up will fail with the standard `KeyError` I suppose. I decided to use the pokemon catch to make sure it always fallback to the old approach on any error, and the user sometimes needs to check the worker log to figure out which one is used.

nicholasturner1 replied:
Ok. There might be a small danger if a user assumes that their containers die because of the health check when it's really because they were idle, but I'm ok with this behavior if you are.

ranlu replied:
I added another commit to check if the worker image has builtin health check, and print a message to slack. Can you review that one as well.
It is harder to notify the user the cause when the workers are killed, in order to do that we need to propagate the error from the docker operator back to the manager. Not sure if there is a clean way to do it.
I could not find a python operator in the synaptor dag to insert the `health_check_info` call, so I only modified segmentation dag and chunkflow dag.

nicholasturner1 commented on plugins/custom/docker_custom.py:
```diff
@@ -181,13 +181,23 @@ def _read_task_stats(self):
+            try:
+                 info = self.cli.inspect_container(container=self.container['Id'])
+                 health_check_ok = info['State']['Health']['Status']
+                 if health_check_ok:
```
Is there any content in the health check status? Or does every value other than `None`/`False` mean that things are ok?

ranlu replied:
When I tested it, when there is a health check for the container, `info['State']['Health']['Status']` is either `True` or `False`,  The `Health` key does not exist if the image does not have a health check and the container is not started with user specified health check.

nicholasturner1 replied:
Got it. Sounds reasonable.

nicholasturner1 reviewed (commented):
Looks good overall. My comments are only trying to gauge how much of an assumption you're making in a few cases.

ranlu committed:
Check whether the images have builtin health checks

Right now the bot just send a slack message, in the future we can change
the bot's behavior accordingly

nicholasturner1 commented on common/docker_helper.py:
```diff
@@ -0,0 +1,15 @@
+def health_check_info(image_name):
+    import docker
+    from slack_message import slack_message
```
You put this file in `common` but `slack_message` is in `dags`. Does this import statement work everywhere you'd use it? Maybe this is related to what you mean by changing the bot's behavior later.

ranlu replied:
I think so, airflow sets the `PYTHONPATH` to include `dags` and `plugins` and I manually add `common`. I plan to gradually move the code that are not specific for the dags to `common`, `slack_message` is one of them, but that change is not directly related to healthcheck, should be put in a different pr

nicholasturner1 replied:
That makes sense.

nicholasturner1 reviewed (approved):
Cool feature! I feel an impulse to put this everywhere so we start getting nagged about including health checks in our docker images. Perhaps expressing too many :disappointed:s would be de-motivating though... What changes are you thinking about adding to this later?

ranlu commented:
We can adjust flags based on whether the image has health check, for example the synaptor dag currently disable qos, we can enable qos when there is a health check in the image
